iroh/src/client/blobs/batch.rs

```diff
@@ -0,0 +1,460 @@
+}
+
+/// Options for adding a directory as a collection
+#[derive(Debug, Clone)]
+pub struct AddDirOpts {
+    /// The import mode
+    pub import_mode: ImportMode,
+    /// Whether to preserve the directory name
+    pub wrap: WrapOption,
+    /// Io parallelism
+    pub io_parallelism: usize,
+}
+
+impl Default for AddDirOpts {
+    fn default() -> Self {
+        Self {
+            import_mode: ImportMode::TryReference,
+            wrap: WrapOption::NoWrap,
+            io_parallelism: 4,
+        }
+    }
+}
+
+/// Options for adding a directory as a collection
+#[derive(Debug, Clone)]
+pub struct AddReaderOpts {
+    /// The format of the blob
+    pub format: BlobFormat,
+    /// Size of the chunks to send
+    pub chunk_size: usize,
+}
+
+impl Default for AddReaderOpts {
+    fn default() -> Self {
+        Self {
+            format: BlobFormat::Raw,
+            chunk_size: 1024 * 64,
+        }
+    }
+}
+
+impl Batch {
+    pub(super) fn new(
+        batch: BatchId,
+        rpc: RpcClient,
+        updates: UpdateSink<RpcService, RpcConnection, BatchUpdate>,
+        buffer_size: usize,
+    ) -> Self {
+        let updates = updates.buffer(buffer_size);
+        Self(Arc::new(BatchInner {
+            batch,
+            rpc,
+            updates: updates.into(),
+        }))
+    }
+
+    /// Write a blob by passing bytes.
+    pub async fn add_bytes(&self, bytes: impl Into<Bytes>) -> Result<TempTag> {
+        self.add_bytes_with_opts(bytes, Default::default()).await
+    }
+
+    /// Import a blob from a filesystem path, using the default options.
+    ///
+    /// For more control, use [`Self::add_file_with_opts`].
+    pub async fn add_file(&self, path: PathBuf) -> Result<(TempTag, u64)> {
+        self.add_file_with_opts(path, AddFileOpts::default()).await
+    }
+
+    /// Add a directory as a hashseq in iroh collection format
+    pub async fn add_dir(&self, root: PathBuf) -> Result<TempTag> {
+        self.add_dir_with_opts(root, Default::default()).await
+    }
+
+    /// Write a blob by passing an async reader.
+    ///
+    /// This will use a default chunk size of 64KB, and a format of [BlobFormat::Raw].
+    pub async fn add_reader(
+        &self,
+        reader: impl AsyncRead + Unpin + Send + 'static,
+    ) -> anyhow::Result<TempTag> {
+        self.add_reader_with_opts(reader, Default::default()).await
+    }
+
+    /// Write a blob by passing a stream of bytes.
+    pub async fn add_stream(
+        &self,
+        input: impl Stream<Item = io::Result<Bytes>> + Send + Unpin + 'static,
+    ) -> Result<TempTag> {
+        self.add_stream_with_opts(input, Default::default()).await
+    }
+
+    /// Create a temp tag to protect some content (blob or hashseq) from being deleted.
+    ///
+    /// A typical use case is that you are downloading some data and want to protect it
+    /// from deletion while the download is ongoing, but don't want to protect it permanently
+    /// until the download is completed.
```

Comment:
```suggestion
    /// Creates a temp tag to protect some content (blob or hashseq) from being deleted.
    ///
    /// This is a lower-level API. The other functions in [`Batch`] already create [`TempTag`]s automatically.
    ///
    /// [`TempTag`]s allow you to protect some data from deletion while a download is ongoing,
    /// even if you don't want to protect it permanently.
```

Not super sure what I'd need the `temp_tag` function for? Tried my best to improve the docs so that people don't expect to *need* to call this function to persist their stuff.